server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,208 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.discovery.DiscoveryModule;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+import org.elasticsearch.health.SimpleHealthIndicatorDetails;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+public class StableMasterHealthIndicatorService implements HealthIndicatorService, ClusterStateListener {
+
+    public static final String NAME = "stable_master";
+
+    private final ClusterService clusterService;
+    private final DiscoveryModule discoveryModule;
+    private final MasterHistoryService masterHistoryService;
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+
+    public StableMasterHealthIndicatorService(
+        ClusterService clusterService,
+        DiscoveryModule discoveryModule,
+        MasterHistoryService masterHistoryService
+    ) {
+        this.clusterService = clusterService;
+        this.discoveryModule = discoveryModule;
+        this.masterHistoryService = masterHistoryService;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean includeDetails) {
+        HealthStatus stableMasterStatus;
+        String summary;
+        Map<String, Object> details = new HashMap<>();
+        Collection<HealthIndicatorImpact> impacts = new ArrayList<>();
+        MasterHistory localMasterHistory = masterHistoryService.getLocalMasterHistory();
+        if (hasSeenMasterInLast30Seconds()) {
+            long masterChanges = getNumberOfMasterChanges(localMasterHistory);
+            logger.trace("Have seen a master in the last 30 seconds");
```

Comment:
Should this print the seen master? Not sure how useful it is in the current form?

Reply:
I'm confused -- the current ActionListener is doing exactly what your proposed new ActionListener would do, right?

Reply:
The previous comment was accidentally in response to the ActionListener thread on the MasterHistoryService class. I didn't notice until just now that I can't respond to that thread for some reason.

> Should this print the seen master? Not sure how useful it is in the current form?

It's useful to see which branch of the code is being taken. But I can throw in the master name.